Database implementation

sebastianfrelle opened the pull request:
For reviewing changes to branch `db` during the implementation of a database.

sirupsen committed:
db: initial template

sebastianfrelle committed:
master merged into db

sebastianfrelle commented on table_foosball/table_foosball_01.rb:
```diff
@@ -44,18 +45,17 @@ def most_losses_against
+    def games
+      @games ||= database.load
+    end
+    alias_method :all, :games
```
Is this just to avoid breaking the tests we've written so far?

sirupsen replied:
yup

sebastianfrelle committed:
serialization of objects implemented with class name and instance variables

sebastianfrelle committed:
serializing objects implemented w/ recursion

Also, test.* added to .gitignore

sebastianfrelle committed:
gitignore stuff

sebastianfrelle committed:
recursion in serialize_objects modified

sebastianfrelle committed:
new serialization syntax (klass_name:value)

sebastianfrelle committed:
serialization DONE

sebastianfrelle committed:
Deserialization with strings for instance variables complete.

- Probably rewriting serialization next to account for Fixnums, Time, etc.
- Some refactoring is probably needed. Initializing Game, Player w/ hashes might not be optimal.

sebastianfrelle committed:
Squashed commits from local branch db-sandbox.
Squashed commit of the following:

commit 0f88d4bc2d3fd23f553681905d839ab259f46111
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Sat May 21 21:26:14 2016 +0200

    Deserialization with strings for instance variables complete.

    - Probably rewriting serialization next to account for Fixnums, Time, etc.
    - Some refactoring is probably needed. Initializing Game, Player w/ hashes might not be optimal.
    - Slight troubleshooting on deserialization. Token games added to games.sdb.

commit f462d61d8c017ceaa895e204344c83711dc7b1fb
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Sat May 21 18:16:18 2016 +0200

    working on bug

commit 4fd74bd8917f8eee769bfe82dd20831143f19f04
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Sat May 21 18:15:54 2016 +0200

    working on bug

commit e05c184f468be159562a9fcd79b6b0d219afbb61
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Sat May 21 16:26:00 2016 +0200

    at initialization now

commit 9f4b90d732004f0352da9b938d6415b56c756654
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Sat May 21 16:25:34 2016 +0200

    at initialization now

commit 54a115d4d347dfbd119b5141e403bf575b7bd609
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Sat May 21 00:59:32 2016 +0200

    serailization added; commencing integration testing

commit 0c4f99cf8829bd8e9f134216537477e0a9613dd1
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Sat May 21 00:49:55 2016 +0200

    BROKEN no serialization. Intermediary commit to fix this

commit b3a8d6443cd0e43cd17d5371d8e752cf96d66fd5
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Sat May 21 00:17:42 2016 +0200

    stash conflict resolved

commit a5f92c46214003256ac7df5f89e06cbb1cc3c216
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Fri May 20 23:36:50 2016 +0200

    deserialization underway

commit f84e35b07c66413aa654219b6706790ebadd726c
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Thu May 19 21:59:33 2016 +0200

    intermediary commit

commit 7e299fc9b14563da95a5662868fe1e6a884a54fb
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Sun May 15 22:28:39 2016 +0200

    serialization DONE

commit 22f6ad686ae0be69559ab9af09e29c4fe46e107f
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Sun May 15 17:57:08 2016 +0200

    intermediary commit

commit 5f68c602fe1c372f4ff33e9159e41b8a395f5526
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Sun May 15 17:54:57 2016 +0200

    intermediary commit

commit 6ef46bade44504edebc68f0389828a35aec1e267
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Thu May 12 18:27:05 2016 +0200

    testing some things

commit 0a7038838d5380aa000cbaae55d33a3bdbdda8f4
Author: Sebastian Frelle Koch <sebastian.frelle@gmail.com>
Date:   Thu May 12 18:26:50 2016 +0200

    testing some things

sebastianfrelle committed:
serialization, deserialization works for Game class. Refactoring probably up next.

Slight touch-ups (comments deleted et.al.)

sebastianfrelle committed:
serialization, deserialization works for Game class. Refactoring probably up next.

sebastianfrelle committed:
Merge branch 'db' of https://github.com/SebastianFrelle/sebastianscripts into db

sebastianfrelle committed:
touchups

sebastianfrelle committed:
unit testing database commenced

sebastianfrelle committed:
added testing for read/write operations on database

sebastianfrelle committed:
testing for database initialization and read/write done; considering testing  module in separate test suite

sebastianfrelle committed:
testing for database initialization and read/write done; considering testing \`TextTools\` module in separate test suite

sebastianfrelle committed:
touchups

sebastianfrelle committed:
compare_object_states in database_test.rb cleaned up

sebastianfrelle committed:
touchups

sebastianfrelle committed:
.gitignore modified

sebastianfrelle committed:
.gitignore modified

sebastianfrelle committed:
db merged into cli

sebastianfrelle committed:
Merge branch 'cli' of https://github.com/SebastianFrelle/sebastianscripts into cli

sebastianfrelle committed:
refactoring commit. Moved database functionality from Game, Player into module Persistence

sebastianfrelle committed:
refactoring commit. Moved database functionality from Game, Player into module Persistence

sebastianfrelle committed:
creating players and games implemented. games at this point is interactive

sebastianfrelle committed:
creation and selection implemented for games and players

sebastianfrelle committed:
deletion of games and players via cli implemented

sebastianfrelle committed:
deletion of games and players via cli implemented

sebastianfrelle committed:
Merge branch 'cli' of https://github.com/SebastianFrelle/sebastianscripts into cli

sebastianfrelle committed:
clear_games takes no options now

sebastianfrelle committed:
Merge branch 'cli' into db

SebastianFrelle commented:
PR #4 (*Command line interface*) merged here

sirupsen commented on table_foosball/database_test.rb:
```diff
@@ -0,0 +1,79 @@
+		
+		assert_equal Database, test_database.class
+		assert_equal name, test_database.instance_eval { @name }
+		assert_equal "./#{name}.sdb", test_database.instance_eval { @filename }
```
This is pretty sketchy. In general, in Ruby the tests tend to be too tightly coupled to the code because you can do so much magic. This magic is something to be weary of as much in tests as in code. If you write test code you would hesitate to write outside of tests, it's usually a good smell that something wrong is going on.

sirupsen commented on table_foosball/database_test.rb:
```diff
@@ -0,0 +1,79 @@
+require_relative './database'
```
This file should have 2-space indentation. This uses tabs.

Sirupsen commented on table_foosball/database_test.rb:
```diff
@@ -0,0 +1,79 @@
+class DatabaseTest < Minitest::Test
+	def setup
+		File.open("./games_test.sdb", "w") {}
+		File.open("./players_test.sdb", "w") {}
```
This is to just create them as empty files?

sirupsen commented on table_foosball/database_test.rb:
```diff
@@ -0,0 +1,79 @@
+
+	private
+
+	def compare_object_states exp, act
```
I see what you're trying to do here. What you ran into is that two objects can't be compared cleanly, e.g. if I do:

``` ruby
class Foo
  def initialize(bar)
    @bar = bar
  end
end
```

And then do:

``` ruby
Foo.new("a") == Foo.new("a")
# => false
```

This is because the default `==` method on objects will only look for complete equality, but you could define your own.

I think your helper here is actually the most future-proof way of doing this, but if each object had an equality method, you wouldn't need this:

``` ruby
class Foo
  attr_reader :bar

  def initialize(bar)
    @bar = bar
  end

  def ==(other)
    other.bar == self.bar
  end
end
```

In which case it'd work, e.g. `foo1 = Foo.new("a"); foo2 = Foo.new("a"); [foo1] == [foo2]` would equal true.

sirupsen commented on table_foosball/database_test.rb:
```diff
@@ -0,0 +1,79 @@
+require 'minitest/autorun'
+
+class DatabaseTest < Minitest::Test
+	def setup
```
I think you should decouple these tests from Games and Players. Instead, define some custom classes directly in here and use them, e.g.

``` ruby
class DatabaseTest < Minitest::Test
  class Foo
    def initialize(..)
    end
  end

  def setup
    # stuff
  end

  # tests
end
```

sirupsen replied:
This means that down the line we can have your database as a completely separate project and potentially use it for other things.

sirupsen commented on table_foosball/persistence.rb:
```diff
@@ -0,0 +1,24 @@
+module Persistence
```
I applaud your exploration of mixins here, but I think this abstraction is unhelpful.

sirupsen commented on table_foosball/table_foosball_01.rb:
```diff
@@ -1,14 +1,39 @@
+    end
+
+    def by_name(name)
+      players.each do |player|
```
There's actually a convenience method for this!

``` ruby
players.find { |player| player.name.downcase == name.downcase }
```

sirupsen commented on table_foosball/database.rb:
```diff
@@ -0,0 +1,117 @@
+        i += 1
+      end
+
+      deserialized_obj = Kernel.const_get(klass_name).new(variables)
```
This requires that what you're dumping takes an initialization hash, which is not very robust. Instead, use the write-equivalent of `instance_variable_get` and use `instance_variable_set`.

sebastianfrelle replied:
I'm still having trouble figuring this one out. We would need to initialize the object and _then_ set its instance variables, correct? We'd need `#initialize` to take `*args`, call `new` with no parameters and then set the instance variables iteratively, which would require `#initialize` for what we're dumping to follow this template, or one that's functionally similar:

``` ruby
def initialize(*args)
    @side1 = args[0]
    @side2 = args[1]
    @side1score = args[2]
    # etc.
end
```

...which doesn't seem that robust either. Am I completely off-track here?

sirupsen replied:
I see what you mean. If you read the implementation of [`Object#new`](https://github.com/ruby/ruby/blob/trunk/object.c#L1852), you'll see that all it does is call `#allocate` and `#initialize`. You could abuse this information to avoid calling `#initialize` by doing this:

``` ruby
object_klass = Kernel.const_get(klass_name)
object = object_klass.allocate
variables.each do |name, value|
  object.instance_variable_set("@#{name}", value)
end
```

sirupsen commented on table_foosball/table_foosball_01.rb:
```diff
@@ -1,14 +1,39 @@
+  def initialize(*args)
+    case args.first.class.name
+    when 'Hash'
+      args[0].each { |k, v| send("#{k}=", v) }
```
Once you use `instance_variable_set`, you won't need this.

sirupsen commented on table_foosball/text_tools.rb:
```diff
@@ -0,0 +1,51 @@
+module TextTools
```
Please 2-space indent this as well.

sirupsen commented on table_foosball/database.rb:
```diff
@@ -0,0 +1,117 @@
+
+  def save(objs)
+    file = File.open(@filename, "r+")    
+    file.rewind
```
So this is kind of interesting. Can you explain what the difference between `File.open(@filename, "w")` and `File.open(@filename, "r+); file.rewind` is?

sebastianfrelle replied:
`"r+"` doesn't truncate the file to length 0 like `"w"` does. The file should be completely overwritten for each call to `#save`, I'd say, so I'll revert back to `"w"`. Thanks for pointing it out!

sirupsen replied:
Exactly! Here's another nifty thing though, let's say you had one process write to this database with another one reading it at the same time. If you truncate the file, that operation doesn't apply instantly (atomicly). That means the file could potentially be corrupt (not valid) when the reader tries to access it. `File.rename` is atomic (instant), but truncating the file is not. That means the way (on Linux) to atomically overwrite a file is to write to a new file, and then replace it with the other one with a rename. Something like this:

``` ruby
File.open("#{@filename}.new", "w") { |f| f.write(serialize_objects(objs)) }
File.rename("#{@filename}.new", @filename)
```

You don't actually have to do this, but thought you might find it interesting :)

sirupsen commented on table_foosball/database.rb:
```diff
@@ -0,0 +1,117 @@
+
+    serialized_objects = file.read
+
+    return [] if serialized_objects == ""
```
Why an empty array and not `nil`? What if you're only serializing an object? Or a Hash? What if you serialized `nil`?

sirupsen commented on table_foosball/text_tools.rb:
```diff
@@ -0,0 +1,51 @@
+
+	end
+
+	def find_next_object(strings, index)
```
Why did you decide to put these in a different module? They belong perfectly inside of the `Database` class as private methods. Usually you'd use a mixin if it's reuseable and can be tested independently, but in this case you're doing neither. This makes it a premature optimization, which I think you should avoid to gain simplicity.

sirupsen commented on table_foosball/database.rb:
```diff
@@ -0,0 +1,117 @@
+require_relative "./text_tools"
+
+class Database
```
I ran into this problem when I was testing the database on my own:

``` ruby
irb(main):006:0> db.save(["a", "b", ["c", "d"]])
=> nil
irb(main):007:0> db.load
=> [nil]
```

sebastianfrelle committed:
de/serialization modified to accomodate nested arrays, database testing is up and running, many other things

sebastianfrelle committed:
small changes

sebastianfrelle committed:
testing rewritten for database; database still broken

sebastianfrelle committed:
serialization done; now supports hashes. database2.rb is where the magic happens.

sebastianfrelle committed:
deserialization of hashes underway

sebastianfrelle committed:
deserialization of objects underway

sebastianfrelle committed:
de/serialization of nested hashes underway. Syntax change upcoming.

sebastianfrelle committed:
de/serialization of nested hashes done. Commencing testing

sebastianfrelle committed:
Testing done for now; database appears to be working as expected.

	- Refactoring and slight cleanup done on database
	- Additional tests written and performed
	- De/serialization of Time objects changed to use rational numbers because of differing precisions and comparisons.
	- Testing class Foo implemented with `#hash` to account for comparing hashes in testing.
	- and many other things.

Next up: cleanup of database code, particularly with regards to the ranges being used in selecting arrays.

Sirupsen commented:
bump with a comment when you're ready for another review

sebastianfrelle committed:
cleanup

SebastianFrelle commented:
Bump!

sebastianfrelle commented on table_foosball/database_test.rb:
```diff
@@ -0,0 +1,139 @@
+
+    alias eql? ==
+
+    def hash
```
So I defined `Foo#hash` to be able to compare hashes containing objects with instance variables as their keys/values. My theory is that `Hash#==` compares objects using their hash values (`#hash`), as hashes containing objects with identical variables would come back as not-equal when I wrote these tests. However, the docs on `Hash#==` tell me that it just makes use of `Object#==`, which I thought I'd overwritten by defining `Foo#==`.

Am I on the right track here? And how do you (@Sirupsen) search Ruby's source code for method definitions?

Edit: this is where I got the idea: http://javieracero.com/blog/the-key-to-ruby-hashes-is-eql-hash

sirupsen replied:
You can't test for equality of objects this way, and providing the class with it is sketchy IMO. I think you should just have a method that does what you're doing in this method:

``` ruby
def assert_object_state_equal?(a, b)
  assert_equal instance_variable_values(a), instance_variable_values(b)
end

def instance_variable_values(obj)
  obj.instance_variables.map { |name| obj.instance_variable_get(name) }
end
```

Sirupsen commented:
I haven't combed finely through this, but I think that's fine :) Merge this and move on, fixing more bugs as you find them.

sebastianfrelle committed:
Changed regex anchor

Changed anchor from ^ to \A for regex check to check start of string. Not tested yet!
